bpcosmo/pm.py

```diff
@@ -0,0 +1,145 @@
+                   density_plane_npix=256,         # Number of pixels
+                   density_plane_smoothing=3.,     # In Mpc/h
+                   box_size=[400.,400.,4000.],     # In Mpc/h
+                   nc=[64,64,640],
+                   neural_spline_params=None
+                   ):
+  """Function that returns tomographic density planes 
+  for a given cosmology from a lightcone.
+
+  Args:
+    cosmology: jax-cosmo object
+    density_plane_width: width of the output density slices
+    density_plane_npix: size of the output density slices
+    density_plane_smoothing: Gaussian scale of plane smoothing
+    box_size: [sx,sy,sz] size in Mpc/h of the simulation volume
+    nc: number of particles/voxels in the PM scheme
+    neural_spline_params: optional parameters for neural correction of PM scheme
+  Returns:
+    list of [r, a, plane], slices through the lightcone along with their
+        comoving distance and scale factors.
+  """
+  # Planning out the scale factor stepping to extract desired lensplanes
+  n_lens = int(box_size[-1] // density_plane_width)
+  r = jnp.linspace(0., box_size[-1], n_lens+1)
+  r_center = 0.5*(r[1:] + r[:-1])
+  a_center = jc.background.a_of_chi(cosmology, r_center)
+
+  # Create a small function to generate the matter power spectrum
+  k = jnp.logspace(-4, 1, 128)
+  pk = jc.power.linear_matter_power(cosmology, k)
+  pk_fn = lambda x: jc.scipy.interpolate.interp(x.reshape([-1]), k, pk).reshape(x.shape)
+
+  # Create initial conditions
+  initial_conditions = linear_field(nc, box_size, pk_fn)
+
+  # Create particles
+  particles = jnp.stack(jnp.meshgrid(*[jnp.arange(s) for s in nc]),axis=-1).reshape([-1,3])
+
+  # Initial displacement
+  cosmology._workspace = {} # FIX ME: this a temporary fix
+  dx, p, f = lpt(cosmology, initial_conditions, particles, 0.01)
+
+  @jax.jit
+  def neural_nbody_ode(state, a, cosmo, params):
+      """
+      state is a tuple (position, velocities)
+      """
+      pos, vel = state
+      
+      kvec = fftk(nc)
+
+      delta = cic_paint(jnp.zeros(nc), pos)
+      
+      delta_k = jnp.fft.rfftn(delta)
+      
+      # Computes gravitational potential
+      pot_k = delta_k * laplace_kernel(kvec) * longrange_kernel(kvec, r_split=0)
+      
+      # Apply a correction filter
+      if params is not None:
+        kk = jnp.sqrt(sum((ki/jnp.pi)**2 for ki in kvec))
+        pot_k = pot_k *(1. + model.apply(params, kk, jnp.atleast_1d(a)))
+      
+      # Computes gravitational forces
+      forces = jnp.stack([cic_read(jnp.fft.irfftn(gradient_kernel(kvec, i)*pot_k), pos) 
+                        for i in range(3)],axis=-1)
+      
+      forces = forces * 1.5 * cosmo.Omega_m
+
+      # Computes the update of position (drift)
+      dpos = 1. / (a**3 * jnp.sqrt(jc.background.Esqr(cosmo, a))) * vel
+      
+      # Computes the update of velocity (kick)
+      dvel = 1. / (a**2 * jnp.sqrt(jc.background.Esqr(cosmo, a))) * forces
```

Comment:
could reuse the scaling factor (not really important though)

Reply:
1. Do we need a reference cited for the neural spline? 
2. Being pedantic here, but maybe it's worth clarifying: "list of [r, a, plane], slices through the lightcone along with their comoving distance (r) and scale factors (a), with 'plane' being the index of the density slice.", or something along those lines.
3. plane should range from 0 to nc[0]?
4. line 72: why 128 bins?
5. line 89: velocities in km/s?
6. line 102: why do we need this correction? what filter is this?

Reply:
yeah but it's just one multiplication and you gain in clarity